File: Sorting_2/mergeSort.js
const mergeSort = (arr, low, high) => {
    if (low >= high) {
        return;
    }
    let mid = Math.floor((low + high) / 2);
    mergeSort(arr, low, mid);
    mergeSort(arr, mid + 1, high);
    merge(arr, low, mid, high);

}
const merge = (arr, low, mid, high) => {
    let left = low;
    let right = mid+1;
    let temp = [];
    while (left <= mid && right <= high) {
        if (arr[left] < arr[right]) {
            temp.push(arr[left]);
            left++;
        } else {
            temp.push(arr[right]);
            right++;
        }
    }
    while (left <= mid) {
        temp.push(arr[left]);
        left++;
    }
    while (right <= high) {
        temp.push(arr[right]);
        right++;
    }
    for(let i=low;i<=high;i++){
        arr[i]=temp[i-low];
    }


}
let a = [21, 24, 5, 6, 7, 8, 67,82];
mergeSort(a, 0, a.length - 1);
console.log(a.join(" "));

File: Sorting_2/mergeSort.java
import java.util.*;

class MergeSort {
    public static void main(String[] args) {
        int[] a = { 1, 21, 4, 12, 5, 80, 61, 84, 65,77 };
        mergeSort(a, 0, a.length - 1);
        for (int i = 0; i < a.length; i++) {

            System.out.print(a[i] + " ");
        }

    }

    public static void mergeSort(int[] arr, int low, int high) {
        int n = arr.length;
        if (low >= high) {
            return;
        }

        int mid = (low + high) / 2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid + 1, high);
        merge(arr, low, mid, high);

    }

    public static void merge(int[] arr, int low, int mid, int high) {
        int left = low;
        int right = mid + 1;
        ArrayList<Integer> temp = new ArrayList<>();
        while (left <= mid && right <= high) {
            if (arr[left] < arr[right]) {
                temp.add(arr[left]);
                left++;
            } else {
                temp.add(arr[right]);
                right++;
            }
        }
        while (left <= mid) {
            temp.add(arr[left]);
            left++;

        }
        while (right <= high) {
            temp.add(arr[right]);
            right++;
        }
        for(int i=low;i<=high;i++){
            arr[i]=temp.get(i-low);
        }
    }
}